### PR TITLE
[SYCL][Graph] Only get command handle for updatable command-buffers

### PR DESCRIPTION
**Pull request by Bensuo**
- Update UR tag to include changes to command-handles
- Update calls to UrCommandBufferAppendKernelLaunch to only request command handles for updatable command-buffers
- Only store command handles for updatable command-graphs

UR PR: https://github.com/oneapi-src/unified-runtime/pull/2131

### CODE REVIEW

File: sycl/source/detail/scheduler/commands.cpp [hunk 1]
```diff
@@ -2535,12 +2535,19 @@ ur_result_t enqueueImpCommandBufferKernel(
       LocalSize = RequiredWGSize;
   }
 
+  ur_exp_command_buffer_desc_t CommandBufferDesc{};
+
+  Adapter->call<UrApiKind::urCommandBufferGetInfoExp>(
```
Comment (EwanC) [inline sycl/source/detail/scheduler/commands.cpp:2540]: Would stick in a wee single line comment saying a handles are only valid when a graph is updatable, to save somebody having to go lookup the UR spec.

File: sycl/source/detail/graph_impl.cpp [hunk 2]
```diff
@@ -676,9 +676,12 @@ exec_graph_impl::enqueueNodeDirect(sycl::context Ctx,
   ur_result_t Res = sycl::detail::enqueueImpCommandBufferKernel(
       Ctx, DeviceImpl, CommandBuffer,
       *static_cast<sycl::detail::CGExecKernel *>((Node->MCommandGroup.get())),
-      Deps, &NewSyncPoint, &NewCommand, nullptr);
+      Deps, &NewSyncPoint, this->MIsUpdatable ? &NewCommand : nullptr, nullptr);
 
-  MCommandMap[Node] = NewCommand;
+  if(this->MIsUpdatable){
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.cpp:681]: nitpick, but can we drop the `this` from these `this->MIsUpdatable` expressions, not sure if it's convention or not.

**Review by EwanC — APPROVED** [review]
LGTM

File: sycl/source/detail/scheduler/commands.cpp (hunk 1 above)
Comment (Bensuo) [reply sycl/source/detail/scheduler/commands.cpp:2542]: Good point, done!

File: sycl/source/detail/graph_impl.cpp (hunk 2 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.cpp]: Removed, it seems like our usage is inconsistent but this is probably bad practice. I've left the other instances alone for now to not clutter this PR with unrelated changes.

**Comment by kbenzie** [comment]
@intel/llvm-reviewers-runtime please review

**Review by kbenzie — APPROVED** [review]

**Comment by steffenlarsen** [comment]
Could you please add a test for this? Could just be a unittest that checks the path.

**Comment by Bensuo** [comment]
> Could you please add a test for this? Could just be a unittest that checks the path.

@steffenlarsen I'm afraid I'm not sure what there is to test on this path that isn't already covered by existing update tests. There is no change in user facing behaviour of graphs, and command handles are stored privately within `exec_graph_impl` so they can't be checked in a unit test.

Was there something specific you had in mind?

**Comment by steffenlarsen** [comment]
You could write a unittest that hits this path and redefines `urCommandBufferAppendKernelLaunchExp` to check the new argument behavior. Mostly just to make sure it's actually hitting that path and continues to do so.

**Comment by kbenzie** [comment]
> You could write a unittest that hits this path and redefines `urCommandBufferAppendKernelLaunchExp` to check the new argument behavior. Mostly just to make sure it's actually hitting that path and continues to do so.

Would it be okay to defer adding a test by creating an issue so we can unblock the UR merge queue?

**Review by steffenlarsen — APPROVED** [review]
Alright, I will budge. Please do remember to add one. 😉

**Comment by steffenlarsen** [comment]
Issue observed with https://github.com/intel/llvm/pull/15309#issuecomment-2421977575 seems to have been reintroduced with this. Tag @nrspruit .

**Comment by aarongreig** [comment]
are we sure the reverted PR was the problem? the trace suggested something was going wrong with urPlatformGet but the init pr doesn't touch anything anywhere near that https://github.com/oneapi-src/unified-runtime/pull/2063/files

**Comment by pbalcer** [comment]
> we had a level zero loader bump merged the same day https://github.com/oneapi-src/unified-runtime/pull/2208/files#diff-d5a56100f25f4eb20d6b4d3ee09aef049a40ec09c6b9fbfe430be22360aaa11f I'm guessing we expected that not to affect this repo since it didn't have a corresponding PR but maybe it did somehow?

e2e tests were passing normally with the loader bump. But it's not impossible that it's environment related.

**Comment by kbenzie** [comment]
Since this is a post-commit issue, how can we run post-commit on a PR to iterate potentialy solutions to this?

**Comment by aarongreig** [comment]
as I understand it there isn't a problem in post-commit, it's something that's only showing up in local builds (I'm seeing it as well fwiw)

**Comment by pbalcer** [comment]
> as I understand it there isn't a problem in post-commit, it's something that's only showing up in local builds (I'm seeing it as well fwiw)

can you try removing any libze* files from the lib directory and see whether this still reproduces? that would definitively confirm that this is an L0 loader issue.

**Comment by aarongreig** [comment]
> > as I understand it there isn't a problem in post-commit, it's something that's only showing up in local builds (I'm seeing it as well fwiw)
> 
> can you try removing any libze* files from the lib directory and see whether this still reproduces? that would definitively confirm that this is an L0 loader issue.

Yeah that fixed it, and now I'm finding that I can't provoke a build into reproducing. Maybe this change just necessitated a clean build to work properly

**Comment by pbalcer** [comment]
> > > as I understand it there isn't a problem in post-commit, it's something that's only showing up in local builds (I'm seeing it as well fwiw)
> > 
> > 
> > can you try removing any libze* files from the lib directory and see whether this still reproduces? that would definitively confirm that this is an L0 loader issue.
> 
> Yeah that fixed it, and now I'm finding that I can't provoke a build into reproducing. Maybe this change just necessitated a clean build to work properly

That would track with the fact that it only reproduced on local builds. It's possible that there were some leftover .so libraries that level zero loader was pulling in.

**Comment by kbenzie** [comment]
Sounds like we should unrevert https://github.com/intel/llvm/pull/15309 and recommend clean builds then?

**Comment by EwanC** [comment]
I'm still seeing the error discussed here in clean build of DPC++  when running sycl-ls:

> SYCL Exception encountered: Native API failed. Native API returns: 2147483646 (UR_RESULT_ERROR_UNKNOWN)

Working around it locally by undoing the change from https://github.com/oneapi-src/unified-runtime/pull/2208 back to `v1.17.39` in the `<build>/_deps/unified-runtime-src/cmake/FetchLevelZero.cmake` file but it's not ideal. I installed the latest L0 drivers from https://github.com/intel/compute-runtime/releases but that didn't help (i'm on a i9-13900K).

**Comment by steffenlarsen** [comment]
Like @EwanC, I am also observing it with a clean build. @kbenzie | @pbalcer - Is anyone investigating this?

**Comment by pbalcer** [comment]
This is like a compute runtime/loader issue. We will file a separate issue, but in the meantime I think we may need to revert the loader change.

**Comment by pbalcer** [comment]
https://github.com/oneapi-src/unified-runtime/pull/2238 created a revert PR here.

**Comment by kbenzie** [comment]
> Like @EwanC, I am also observing it with a clean build. @kbenzie | @pbalcer - Is anyone investigating this?

I've opened GSD-10233 to track this internally.

**Comment by kbenzie** [comment]
> [oneapi-src/unified-runtime#2238](https://github.com/oneapi-src/unified-runtime/pull/2238) created a revert PR here.

This has been merged to UR main and will be pulled to the sycl branch in as part of https://github.com/intel/llvm/pull/15800

**Pull request merged**